File: inkeep_core/pow.py
import hashlib
import json
import base64

class PoWSolver:
    @staticmethod
    def solve(challenge_data):
        """
        Solves the Inkeep Altcha PoW challenge.
        Algorithm: SHA-256(salt + str(number)) == challenge
        """
        challenge = challenge_data.get('challenge')
        salt = challenge_data.get('salt')
        max_number = challenge_data.get('maxnumber', 100000)
        signature = challenge_data.get('signature')

        if not challenge or not salt or not signature:
            raise ValueError("Invalid challenge data")

        for i in range(max_number + 1):
            h = hashlib.sha256((salt + str(i)).encode()).hexdigest()
            if h == challenge:
                solution = {
                    "number": i,
                    "algorithm": "SHA-256",
                    "challenge": challenge,
                    "maxnumber": max_number,
                    "salt": salt,
                    "signature": signature
                }
                return base64.b64encode(json.dumps(solution).encode()).decode()
        
        raise Exception(f"Failed to solve PoW challenge within max_number={max_number}")

File: web/src/lib/inkeep/pow.ts
export interface ChallengeData {
  algorithm: string;
  challenge: string;
  maxnumber: number;
  salt: string;
  signature: string;
}

export interface Solution {
  number: number;
  algorithm: string;
  challenge: string;
  maxnumber: number;
  salt: string;
  signature: string;
}

export async function solveChallenge(challengeData: ChallengeData): Promise<string> {
  const { challenge, salt, maxnumber } = challengeData;
  const encoder = new TextEncoder();
  
  // Brute force from 0 to maxnumber
  for (let i = 0; i <= maxnumber; i++) {
    const candidate = i.toString();
    const data = encoder.encode(salt + candidate);
    
    // Calculate SHA-256
    const hashBuffer = await crypto.subtle.digest('SHA-256', data);
    
    // Convert to hex string
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    const hashHex = hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
    
    if (hashHex === challenge) {
      // Found it! Construct solution JSON
      const solution: Solution = {
        number: i,
        algorithm: "SHA-256",
        challenge: challenge,
        maxnumber: maxnumber,
        salt: salt,
        signature: challengeData.signature
      };
      
      // Return Base64 encoded JSON string
      const jsonStr = JSON.stringify(solution);
      return btoa(jsonStr);
    }
  }
  
  throw new Error(`Failed to solve PoW challenge within max_number=${maxnumber}`);
}